File: build_merkle.py
#!/usr/bin/env python3
"""
Build a deterministic keccak256 Merkle tree from account shards and persist it to SQLite.

The tree is built in the exact order defined by shards/manifest.txt, emitting
progress updates as leaves and levels are processed. At completion, the script
prints the Merkle root and the number of accounts included.
"""

from __future__ import annotations

import argparse
import sqlite3
import sys
import time
from pathlib import Path
from typing import Callable, Iterable, Iterator, List, Optional, Tuple


# --- Minimal pure-Python keccak256 (Ethereum-compatible) --------------------

# Rotation offsets for keccak-f[1600]
_RHO = (
    (0, 36, 3, 41, 18),
    (1, 44, 10, 45, 2),
    (62, 6, 43, 15, 61),
    (28, 55, 25, 21, 56),
    (27, 20, 39, 8, 14),
)

_RC = (
    0x0000000000000001,
    0x0000000000008082,
    0x800000000000808A,
    0x8000000080008000,
    0x000000000000808B,
    0x0000000080000001,
    0x8000000080008081,
    0x8000000000008009,
    0x000000000000008A,
    0x0000000000000088,
    0x0000000080008009,
    0x000000008000000A,
    0x000000008000808B,
    0x800000000000008B,
    0x8000000000008089,
    0x8000000000008003,
    0x8000000000008002,
    0x8000000000000080,
    0x000000000000800A,
    0x800000008000000A,
    0x8000000080008081,
    0x8000000000008080,
    0x0000000080000001,
    0x8000000080008008,
)

_MASK_64 = (1 << 64) - 1


def _rotl64(x: int, n: int) -> int:
    return ((x << n) & _MASK_64) | (x >> (64 - n))


def _keccak_f(state: List[int]) -> None:
    for rc in _RC:
        # Theta
        c = [state[x] ^ state[x + 5] ^ state[x + 10] ^ state[x + 15] ^ state[x + 20] for x in range(5)]
        d = [c[(x - 1) % 5] ^ _rotl64(c[(x + 1) % 5], 1) for x in range(5)]
        for x in range(5):
            for y in range(5):
                state[x + 5 * y] ^= d[x]

        # Rho and Pi
        b = [0] * 25
        for x in range(5):
            for y in range(5):
                b[y + 5 * ((2 * x + 3 * y) % 5)] = _rotl64(state[x + 5 * y], _RHO[x][y])

        # Chi
        for x in range(5):
            for y in range(5):
                state[x + 5 * y] = b[x + 5 * y] ^ ((~b[(x + 1) % 5 + 5 * y]) & b[(x + 2) % 5 + 5 * y])

        # Iota
        state[0] ^= rc


def keccak256_digest(data: bytes) -> bytes:
    """Keccak-256 with pad10*1 (Ethereum style)."""
    rate = 136  # bytes
    state = [0] * 25  # 5x5 lanes of 64 bits
    offset = 0
    # Absorb full blocks
    while offset + rate <= len(data):
        block = data[offset : offset + rate]
        for i in range(rate // 8):
            lane = int.from_bytes(block[8 * i : 8 * (i + 1)], "little")
            state[i] ^= lane
        _keccak_f(state)
        offset += rate

    # Pad remaining bytes with 0x01 ... 0x80
    remaining = data[offset:]
    padded = bytearray(rate)
    padded[: len(remaining)] = remaining
    padded[len(remaining)] = 0x01
    padded[-1] |= 0x80
    for i in range(rate // 8):
        lane = int.from_bytes(padded[8 * i : 8 * (i + 1)], "little")
        state[i] ^= lane
    _keccak_f(state)

    # Squeeze digest
    out = bytearray()
    while len(out) < 32:
        for i in range(rate // 8):
            out.extend(state[i].to_bytes(8, "little"))
            if len(out) >= 32:
                break
        if len(out) >= 32:
            break
        _keccak_f(state)
    return bytes(out[:32])


def _resolve_keccak256() -> Tuple[str, Callable[[bytes], bytes]]:
    """
    Prefer fast keccak256 implementations if available, else fall back to the
    built-in pure Python version.
    """
    try:
        from sha3 import keccak_256  # type: ignore

        def digest(data: bytes) -> bytes:
            h = keccak_256()
            h.update(data)
            return h.digest()

        return "keccak256-pysha3", digest
    except Exception:
        pass

    try:
        # pycryptodomex (Cryptodome namespace)
        from Cryptodome.Hash import keccak  # type: ignore

        def digest(data: bytes) -> bytes:
            k = keccak.new(digest_bits=256)
            k.update(data)
            return k.digest()

        return "keccak256-pycryptodomex", digest
    except Exception:
        pass

    try:
        # pycryptodome (Crypto namespace) — less likely here but try.
        from Crypto.Hash import keccak  # type: ignore

        def digest(data: bytes) -> bytes:
            k = keccak.new(digest_bits=256)
            k.update(data)
            return k.digest()

        return "keccak256-pycryptodomex", digest
    except Exception:
        pass

    return "keccak256-purepython", keccak256_digest


# Deterministic hash function used for leaves and nodes (keccak256).
HASH_NAME, HASH_FUNC = _resolve_keccak256()


# --- Merkle construction ----------------------------------------------------


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Create a deterministic Merkle tree from shard files and store it in SQLite."
        )
    )
    parser.add_argument(
        "--shard-dir",
        default="shards",
        help="Directory containing shard files and manifest.txt (default: shards)",
    )
    parser.add_argument(
        "--manifest",
        default=None,
        help="Manifest file listing shard filenames in order (default: <shard-dir>/manifest.txt)",
    )
    parser.add_argument(
        "--database",
        default="merkle.db",
        help="SQLite database file to store the tree (default: merkle.db)",
    )
    parser.add_argument(
        "--report-every",
        type=float,
        default=5.0,
        metavar="SECONDS",
        help="Seconds between progress updates (default: 5.0)",
    )
    parser.add_argument(
        "--batch-size",
        type=int,
        default=50_000,
        help="Number of nodes to insert per transaction (default: 50,000)",
    )
    return parser.parse_args()


def load_manifest(manifest_path: Path) -> List[Path]:
    if not manifest_path.exists():
        sys.exit(f"Manifest not found: {manifest_path}")
    shard_files: List[Path] = []
    with open(manifest_path, "r", encoding="utf-8") as mf:
        for line in mf:
            name = line.strip()
            if not name:
                continue
            shard_files.append(manifest_path.parent / name)
    if not shard_files:
        sys.exit(f"No shard entries found in manifest: {manifest_path}")
    return shard_files


def hash_leaf(address: str) -> bytes:
    addr = address.strip()
    if addr.startswith("0x") or addr.startswith("0X"):
        addr = addr[2:]
    try:
        raw = bytes.fromhex(addr)
    except ValueError as exc:
        raise ValueError(f"Invalid hex address: {address!r}") from exc
    if len(raw) != 20:
        raise ValueError(f"Address must be 20 bytes: {address!r}")
    return HASH_FUNC(raw)


def hash_pair(left: bytes, right: bytes) -> bytes:
    return HASH_FUNC(left + right)


def init_db(db_path: Path) -> sqlite3.Connection:
    conn = sqlite3.connect(db_path)
    # Speed-oriented pragmas: disable journaling for this one-shot build, in-memory temp.
    conn.execute("PRAGMA journal_mode=OFF;")
    conn.execute("PRAGMA synchronous=OFF;")
    conn.execute("PRAGMA temp_store=MEMORY;")
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS nodes (
            level INTEGER NOT NULL,
            idx   INTEGER NOT NULL,
            hash  BLOB NOT NULL,
            PRIMARY KEY (level, idx)
        ) WITHOUT ROWID;
        """
    )
    conn.execute(
        """
        CREATE TABLE IF NOT EXISTS meta (
            key TEXT PRIMARY KEY,
            value TEXT NOT NULL
        ) WITHOUT ROWID;
        """
    )
    return conn


def store_meta(conn: sqlite3.Connection, key: str, value: str) -> None:
    conn.execute(
        "INSERT OR REPLACE INTO meta (key, value) VALUES (?, ?)",
        (key, value),
    )


def read_addresses(shard_files: Iterable[Path]) -> Iterator[str]:
    for shard in shard_files:
        if not shard.exists():
            raise FileNotFoundError(f"Shard file missing: {shard}")
        with open(shard, "r", encoding="utf-8") as fh:
            for line in fh:
                addr = line.strip()
                if not addr:
                    continue
                yield addr


def insert_batch(
    conn: sqlite3.Connection, batch: List[Tuple[int, int, bytes]]
) -> None:
    conn.executemany(
        "INSERT OR REPLACE INTO nodes (level, idx, hash) VALUES (?, ?, ?)", batch
    )


def build_leaves(
    conn: sqlite3.Connection,
    shard_files: List[Path],
    batch_size: int,
    report_every: float,
) -> int:
    conn.execute("DELETE FROM nodes")
    leaf_idx = 0
    batch: List[Tuple[int, int, bytes]] = []
    last_report = time.time()

    print(
        f"Reading shards ({len(shard_files)} files) and hashing leaves...",
        file=sys.stderr,
    )

    for addr in read_addresses(shard_files):
        h = hash_leaf(addr)
        batch.append((0, leaf_idx, h))
        leaf_idx += 1

        if len(batch) >= batch_size:
            insert_batch(conn, batch)
            conn.commit()
            batch.clear()

        now = time.time()
        if now - last_report >= max(0.5, report_every):
            print(f"Leaves processed: {leaf_idx:,}", file=sys.stderr)
            last_report = now

    if batch:
        insert_batch(conn, batch)
        conn.commit()

    print(f"Finished leaves: {leaf_idx:,} accounts", file=sys.stderr)
    return leaf_idx


def iterate_level_hashes(
    conn: sqlite3.Connection, level: int
) -> Iterator[Tuple[int, bytes]]:
    cursor = conn.execute(
        "SELECT idx, hash FROM nodes WHERE level = ? ORDER BY idx ASC", (level,)
    )
    for idx, h in cursor:
        yield int(idx), h


def build_level(
    conn: sqlite3.Connection,
    level: int,
    count: int,
    batch_size: int,
    report_every: float,
) -> int:
    parent_level = level + 1
    batch: List[Tuple[int, int, bytes]] = []
    parent_idx = 0
    last_report = time.time()
    processed = 0

    print(
        f"Building level {parent_level} from {count:,} node(s)...",
        file=sys.stderr,
    )

    pending: Optional[Tuple[int, bytes]] = None
    for idx, h in iterate_level_hashes(conn, level):
        if pending is None:
            pending = (idx, h)
            continue

        left = pending[1]
        right = h
        parent_hash = hash_pair(left, right)
        batch.append((parent_level, parent_idx, parent_hash))
        parent_idx += 1
        processed += 2
        pending = None

        if len(batch) >= batch_size:
            insert_batch(conn, batch)
            conn.commit()
            batch.clear()

        now = time.time()
        if now - last_report >= max(0.5, report_every):
            pct = processed / count * 100 if count else 100.0
            print(
                f"Level {parent_level}: {pct:5.1f}% ({processed:,}/{count:,})",
                file=sys.stderr,
            )
            last_report = now

    if pending is not None:
        # Odd count: duplicate the last hash to keep tree deterministic.
        parent_hash = hash_pair(pending[1], pending[1])
        batch.append((parent_level, parent_idx, parent_hash))
        parent_idx += 1

    if batch:
        insert_batch(conn, batch)
        conn.commit()

    print(
        f"Level {parent_level} complete: {parent_idx:,} node(s)",
        file=sys.stderr,
    )
    return parent_idx


def build_tree(
    conn: sqlite3.Connection,
    leaf_count: int,
    batch_size: int,
    report_every: float,
) -> bytes:
    level = 0
    count = leaf_count
    while count > 1:
        count = build_level(conn, level, count, batch_size, report_every)
        level += 1

    cursor = conn.execute(
        "SELECT hash FROM nodes WHERE level = ? AND idx = 0", (level,)
    )
    row = cursor.fetchone()
    if row is None:
        raise RuntimeError("Failed to compute Merkle root.")
    return row[0]


def main() -> None:
    args = parse_args()
    shard_dir = Path(args.shard_dir)
    manifest_path = Path(args.manifest) if args.manifest else shard_dir / "manifest.txt"
    shard_files = load_manifest(manifest_path)

    conn = init_db(Path(args.database))
    store_meta(conn, "hash_function", HASH_NAME)
    store_meta(conn, "shard_dir", str(shard_dir))
    store_meta(conn, "manifest", str(manifest_path))
    conn.commit()

    print(f"Hash function: {HASH_NAME}", file=sys.stderr)
    start = time.time()
    leaf_count = build_leaves(conn, shard_files, args.batch_size, args.report_every)
    root = build_tree(conn, leaf_count, args.batch_size, args.report_every)
    duration = time.time() - start

    store_meta(conn, "leaf_count", str(leaf_count))
    store_meta(conn, "root_hex", root.hex())
    store_meta(conn, "built_at", time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()))
    conn.commit()
    conn.close()

    print(f"Merkle root: {root.hex()}")
    print(f"Accounts included: {leaf_count:,}")
    print(f"Elapsed: {duration:,.1f}s")


if __name__ == "__main__":
    main()

File: rust_builder/src/build_merkle.rs
use clap::Parser;
use rusqlite::{params, Connection};
use chrono::Utc;
use std::{
    fs::File,
    io::{BufRead, BufReader},
    path::{Path, PathBuf},
    time::{Duration, Instant},
};
use tiny_keccak::{Hasher, Keccak};

#[derive(Parser, Debug)]
#[command(about = "Build a Merkle tree from shards into merkle.db (nodes + meta).")]
struct Args {
    /// SQLite database to create/overwrite
    #[arg(long, default_value = "merkle.db")]
    database: PathBuf,

    /// Directory containing shard files
    #[arg(long, default_value = "shards")]
    shard_dir: PathBuf,

    /// Manifest file listing shard filenames in order
    #[arg(long)]
    manifest: Option<PathBuf>,

    /// Nodes per batch insert/commit
    #[arg(long, default_value_t = 50_000)]
    batch_size: usize,

    /// Seconds between progress logs
    #[arg(long, default_value_t = 5.0)]
    report_every: f64,
}

fn normalize_addr(addr: &str) -> &str {
    let mut a = addr.trim();
    if let Some(stripped) = a.strip_prefix("0x").or_else(|| a.strip_prefix("0X")) {
        a = stripped;
    }
    a
}

fn hash_leaf(addr: &str) -> Result<[u8; 32], String> {
    let norm = normalize_addr(addr);
    let raw = hex::decode(norm).map_err(|e| format!("decode {addr}: {e}"))?;
    if raw.len() != 20 {
        return Err(format!("address must be 20 bytes: {addr}"));
    }
    Ok(keccak256(&raw))
}

fn keccak256(data: &[u8]) -> [u8; 32] {
    let mut hasher = Keccak::v256();
    hasher.update(data);
    let mut out = [0u8; 32];
    hasher.finalize(&mut out);
    out
}

fn hash_pair(left: &[u8], right: &[u8]) -> [u8; 32] {
    let mut buf = Vec::with_capacity(left.len() + right.len());
    buf.extend_from_slice(left);
    buf.extend_from_slice(right);
    keccak256(&buf)
}

fn load_manifest(manifest: &Path) -> Result<Vec<PathBuf>, String> {
    let file = File::open(manifest)
        .map_err(|e| format!("manifest {manifest:?} cannot be opened: {e}"))?;
    let mut shards = Vec::new();
    for line in BufReader::new(file).lines() {
        let line = line.map_err(|e| format!("read manifest line: {e}"))?;
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        shards.push(
            manifest
                .parent()
                .unwrap_or_else(|| Path::new("."))
                .join(trimmed),
        );
    }
    if shards.is_empty() {
        return Err(format!("manifest {manifest:?} has no entries"));
    }
    Ok(shards)
}

fn apply_speed_pragmas(conn: &Connection) -> rusqlite::Result<()> {
    conn.execute_batch(
        r#"
        PRAGMA journal_mode = OFF;
        PRAGMA synchronous = OFF;
        PRAGMA temp_store = MEMORY;
        PRAGMA locking_mode = EXCLUSIVE;
        PRAGMA mmap_size = 1073741824;
        PRAGMA cache_size = -524288;
        PRAGMA foreign_keys = OFF;
    "#,
    )
}

fn ensure_tables(conn: &Connection) -> rusqlite::Result<()> {
    conn.execute_batch(
        r#"
        CREATE TABLE IF NOT EXISTS nodes (
            level INTEGER NOT NULL,
            idx   INTEGER NOT NULL,
            hash  BLOB NOT NULL,
            PRIMARY KEY (level, idx)
        ) WITHOUT ROWID;
        CREATE TABLE IF NOT EXISTS meta (
            key TEXT PRIMARY KEY,
            value TEXT NOT NULL
        ) WITHOUT ROWID;
    "#,
    )
}

fn store_meta(conn: &Connection, key: &str, value: &str) -> rusqlite::Result<()> {
    conn.execute(
        "INSERT OR REPLACE INTO meta (key, value) VALUES (?1, ?2)",
        params![key, value],
    )?;
    Ok(())
}

fn build_leaves(
    conn: &mut Connection,
    shards: &[PathBuf],
    batch_size: usize,
    report_every: Duration,
) -> Result<usize, Box<dyn std::error::Error>> {
    conn.execute("DELETE FROM nodes", [])?;
    let mut inserted: usize = 0;
    let mut batch = Vec::with_capacity(batch_size);
    let start = Instant::now();
    let mut last = start;

    eprintln!(
        "Reading shards ({} files) and hashing leaves...",
        shards.len()
    );
    for shard in shards {
        let file = File::open(shard)?;
        for line in BufReader::new(file).lines() {
            let addr = line?;
            if addr.trim().is_empty() {
                continue;
            }
            let h = hash_leaf(&addr)?;
            batch.push((0_i64, inserted as i64, h));
            inserted += 1;

            if batch.len() >= batch_size {
                flush_batch(conn, &batch)?;
                batch.clear();
            }

            if last.elapsed() >= report_every {
                eprintln!(
                    "Leaves processed: {} | {:.0} rows/s",
                    inserted,
                    inserted as f64 / start.elapsed().as_secs_f64().max(0.001)
                );
                last = Instant::now();
            }
        }
    }

    if !batch.is_empty() {
        flush_batch(conn, &batch)?;
    }

    eprintln!("Finished leaves: {} accounts", inserted);
    Ok(inserted)
}

fn flush_batch(
    conn: &mut Connection,
    batch: &[(i64, i64, [u8; 32])],
) -> rusqlite::Result<()> {
    let tx = conn.transaction()?;
    {
        let mut stmt = tx.prepare_cached(
            "INSERT OR REPLACE INTO nodes (level, idx, hash) VALUES (?1, ?2, ?3)",
        )?;
        for (level, idx, hash) in batch {
            stmt.execute(params![level, idx, &hash[..]])?;
        }
    }
    tx.commit()?;
    Ok(())
}

fn build_level(
    conn: &mut Connection,
    level: i64,
    count: usize,
    batch_size: usize,
    report_every: Duration,
) -> Result<usize, Box<dyn std::error::Error>> {
    let parent_level = level + 1;
    let mut batch = Vec::with_capacity(batch_size);
    let mut parent_idx: usize = 0;
    let mut processed: usize = 0;
    let start = Instant::now();
    let mut last = start;

    eprintln!(
        "Building level {} from {} node(s)...",
        parent_level, count
    );

    let mut pending: Option<[u8; 32]> = None;
    let chunk_size = batch_size.saturating_mul(4).max(10_000);
    let mut offset: usize = 0;
    while offset < count {
        let upper = usize::min(offset + chunk_size, count);
        let mut stmt = conn.prepare(
            "SELECT hash FROM nodes WHERE level = ? AND idx >= ? AND idx < ? ORDER BY idx ASC",
        )?;
        let mut rows = stmt.query(params![level, offset as i64, upper as i64])?;
        while let Some(row) = rows.next()? {
            let hash: Vec<u8> = row.get(0)?;
            let mut arr = [0u8; 32];
            arr.copy_from_slice(&hash);

            if let Some(left) = pending.take() {
                let parent_hash = hash_pair(&left, &arr);
                batch.push((parent_level, parent_idx as i64, parent_hash));
                parent_idx += 1;
                processed += 2;
            } else {
                pending = Some(arr);
            }
        }
        drop(rows);
        drop(stmt);

        if batch.len() >= batch_size {
            flush_batch(conn, &batch)?;
            batch.clear();
        }

        if last.elapsed() >= report_every {
            let pct = processed as f64 / count as f64 * 100.0;
            eprintln!(
                "Level {}: {:5.1}% ({}/{})",
                parent_level, pct, processed, count
            );
            last = Instant::now();
        }

        offset = upper;
    }

    if let Some(last_hash) = pending {
        let dup = hash_pair(&last_hash, &last_hash);
        batch.push((parent_level, parent_idx as i64, dup));
        parent_idx += 1;
    }

    if !batch.is_empty() {
        flush_batch(conn, &batch)?;
    }

    eprintln!("Level {} complete: {} node(s)", parent_level, parent_idx);
    Ok(parent_idx)
}

fn build_tree(
    conn: &mut Connection,
    leaf_count: usize,
    batch_size: usize,
    report_every: Duration,
) -> Result<[u8; 32], Box<dyn std::error::Error>> {
    let mut level: i64 = 0;
    let mut count = leaf_count;
    while count > 1 {
        count = build_level(conn, level, count, batch_size, report_every)?;
        level += 1;
    }

    let root: Vec<u8> = conn.query_row(
        "SELECT hash FROM nodes WHERE level = ? AND idx = 0",
        params![level],
        |row| row.get(0),
    )?;
    let mut out = [0u8; 32];
    out.copy_from_slice(&root);
    Ok(out)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    let manifest_path = args
        .manifest
        .clone()
        .unwrap_or_else(|| args.shard_dir.join("manifest.txt"));
    let shards = load_manifest(&manifest_path)?;

    let report_every = Duration::from_secs_f64(args.report_every.max(0.1));
    let batch_size = args.batch_size.max(1);

    let mut conn = Connection::open(&args.database)?;
    apply_speed_pragmas(&conn)?;
    ensure_tables(&conn)?;
    store_meta(&conn, "hash_function", "keccak256-tiny-keccak")?;
    store_meta(&conn, "shard_dir", &args.shard_dir.to_string_lossy())?;
    store_meta(&conn, "manifest", &manifest_path.to_string_lossy())?;

    eprintln!("Hash function: keccak256-tiny-keccak");
    let start = Instant::now();
    let leaf_count = build_leaves(&mut conn, &shards, batch_size, report_every)?;
    let root = build_tree(&mut conn, leaf_count, batch_size, report_every)?;
    let elapsed = start.elapsed().as_secs_f64();

    store_meta(&conn, "leaf_count", &leaf_count.to_string())?;
    store_meta(&conn, "root_hex", &hex::encode(root))?;
    store_meta(
        &conn,
        "built_at",
        &Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string(),
    )?;
    conn.close().ok();

    println!("Merkle root: {}", hex::encode(root));
    println!("Accounts included: {}", leaf_count);
    println!("Elapsed: {:.1}s", elapsed);
    Ok(())
}
